### PR TITLE
Streaming export API

### PR DESCRIPTION
**Pull request by fsluis**
# About this fork / pull request

## Why this fork?
The export wrapper of the Mailchimp API wrapper "Gibbon" loads all the exported data first in memory, before returning it. Since with exporting data, there can be lots and lots of data, this seemed fairly inefficient. This fork solves this.

## How is this fork different?
The changes are as following:
* Instead of httparty, the basic net::http interface is used to stream data directly from the Mailchimp API.
* The export API functions now require a &block to feed each row of data to.
* The export test functions use webmock to fake and verify net::http requests.
* Some re-arrangement of functions (parse_response at APICategory.rb)

## What can I do with this fork?
All you can do with the normal Gibbon API Wrapper. Plus, parsing each row of exported data while it is being received from the export API:
```
    ge = Gibbon::Export.new("your_api_key")
    ge.list({:id => *list_id*}) { |row| *do_sth_with* row }
```

It is important to supply an explicit block / procedure to the export functions, not an implicit one. So, the preceding and following one will work. Please note this method also includes a counter (*i*, starting at 0) telling which row of data you're receiving:
```
    method = Proc.new do |row, i|
        *do_sth_with* row
    end
    ge.list(params, &method)
```

Please note, the following example gives a block that is outside of the function and therefore **won't** work:
```
    ge.list({:id => *list_id*}) do |row|
        *do_sth_with* row
    end
```

And, the original way of working with gibbon is still functional of course:
```
    list = ge.list({:id => *list_id*})
```

## Some final words
Happy coding!

### CODE REVIEW

File: spec/gibbon/gibbon_spec.rb [hunk 1]
```diff
@@ -251,34 +258,52 @@
...
 
     it "should handle a single empty space response without throwing an exception" do
       @gibbon.throws_exceptions = true
-      allow(Gibbon::Export).to receive(:post).and_return(Struct.new(:body).new(" "))
+      stub_request(:post, @url + 'sayHello/').
+        to_return(:body => " ", :status => 200)
+      #allow(Gibbon::Export).to receive(:post).and_return(Struct.new(:body).new(" "))
 
-      expect(@gibbon.say_hello(@body)).to eq([])
+      expect(@gibbon.say_hello(@body)).to eq([[]])
```
Comment (daveespo) [inline spec/gibbon/gibbon_spec.rb]: What's up with this change? Why would an "empty" response return an array of 1 empty array? See #102 and #113

File: spec/gibbon/gibbon_spec.rb [hunk 2]
```diff
@@ -1,5 +1,7 @@
 require 'spec_helper'
 require 'cgi'
+require 'webmock'
+require 'webmock/rspec'
```
Comment (daveespo) [inline spec/gibbon/gibbon_spec.rb]: Is the introduction of webmock necessary because Net::HTTP can't be mocked using rspec like the other API tests?

**Comment by daveespo** [comment]
It looks like this change is backwards compatible (i.e. returns the array of JSON objects if no block is given) .. correct?

Also, the notes above say that this works:

````
ge.list({:id => *list_id*}) { |row| *do_sth_with* row }
````

while this doesn't:

````
ge.list({:id => *list_id*}) do |row|
     *do_sth_with* row
end
````

Aren't those functionally equivalent?

I'm +1 on the overall spirit of this pull request. I'm a little wary of the edge case handling since it's hard to test but I can't really see anything wrong at a glance

File: lib/gibbon/export.rb [hunk 3]
```diff
@@ -14,28 +14,47 @@ def export_api_url
...
+        http.request req do |response|
+          i = -1
+          last = ''
+          response.read_body do |chunk|
+            #puts "Chunk length: #{chunk.length}"
+            #puts "Chunk: #{chunk}"
+            next if chunk.nil?
+            lines = (last+chunk).split("\n")
+            # There seems to be a bug (?) in the export API. Sometimes there's not a newline in between json objects...
+            lines = split_json(lines, '][')
```
Comment (daveespo) [inline lib/gibbon/export.rb]: Isn't this risky since the body of each line could change one day to contain a more rich JSON object (potentially containing arrays or json objects)? The only spec that Mailchimp publishes says that each JSON object is separated by a new line.

````
Rows are delimited using a newline (\n) marker so implementations can read in a single line at a time, handle it, and move on.
````

**Comment by amro** [comment]
I'm traveling so it'll be a few days before I can look at this. So you know, I'm already working on version 2, which hits MailChimp's API 3.0 instead.

**Comment by amro** [comment]
(And thanks for taking the time!)

**Comment by fsluis** [comment]
Dear David and Amro,

Thank you for your quick replies. I was myself also more busy the past couple of days than expected, luckily now that's better.

In reply to the main points:
- Indeed, I noticed (too late) that you were already working on version 2. The current patch is for version 1, maybe in the future we can port it to version 2. Of course, this depends on the setup of version 2 and how this differs from version 1.
- Yes, this patch is backwards compatible. When no block is supplied, the results will simply be returned as an array, equally to how it was before.
- About the notes. Somehow, and this is a ruby gimmick, these two examples on how to supply a block are not compatible. To my own surprise! Gave me some headache to figure this out... See [here](http://rubylearning.com/satishtalim/ruby_blocks.html) and, for more extensive documentation about blocks, [here](https://innig.net/software/ruby/closures-in-ruby). The key here is to supply an explicit block that can be passed as an argument between functions. Proc.new and {} works for that, simple do ... end does not because it creates an implicit block.

I'll look into the detailed comments next.

File: lib/gibbon/export.rb (hunk 3 above)
Comment (fsluis) [inline lib/gibbon/export.rb]: In theory this can be risky. When someone has the combinations of '][' or '}{' in their email addresses or click-urls (whether that's possible at all I don't know), this might cause a erroneous split. The risky part can be solved by checking whether the back-to-back bracket combinations are within a String or not.

But, it is not pretty for sure! This is a bug @ mailchimp export api, which should be dealt with by them accordingly - not here. For the time being however, this does form a possible workaround.

I see three options:
a) Leave it like this;
b) Add a check that the erroneous json code is not within a string;
c) Remove it.

File: spec/gibbon/gibbon_spec.rb (hunk 2 above)
Comment (fsluis) [inline spec/gibbon/gibbon_spec.rb]: Using rspec to stub net http response to return fake chunks of json did not work (`response.read_body do |chunk|`, export.rb line 33). One can, of course, use rspec to stub the response method to yield a different, fake, HTTPResponse at `http.request req do |response|` (export.rb, line 33). But, to prevent from having to clutter the test file with creating fake HTTPResponse objects, webmock gives an excellent tool that does this for you.

So, strictly speaking, it is not necessary. But it is preferred!

File: spec/gibbon/gibbon_spec.rb (hunk 1 above)
Comment (fsluis) [inline spec/gibbon/gibbon_spec.rb]: Agreed. There is an easy fix for this, I've committed it already to my fork.
In export.rb line 39, change
``next if chunk.nil?``
into
``next if chunk.nil? or chunk.strip.empty?``

And, of course, revert the change to gibbon_spec as shown above.

**Comment by amro** [comment]
Getting back to this. Thanks again for adding streaming support. I'll accept this patch and release it as a final version of 1.x, but I would like some minor changes. You removed a lot of examples, etc. from the README and changed the gemspec to reflect your fork, etc. I'd like the examples put back in the README and your examples added instead. And the changes to the gemspec relating to your fork reverted.

**Comment by fsluis** [comment]
Thank you for accepting. The changes to the gemspec / readme were only for my fork, not for the master branch of course. I've made these updates now:
1. I reverted back gemspec. I still left the version on "Gibbon::VERSION" now, so one can also verify programmatically which version the gem has.
2. I reverted back the readme, and added my examples there.

**Comment by amro** [comment]
Sorry if I'm a bit slow to respond -- I'm traveling. The Gibbon::VERSION change is just fine by me but it seems to be causing a build failure on CI. Could you fix that?

**Comment by fsluis** [comment]
My bad. Fixed.

**Comment by amro** [comment]
NP, thanks for contributing

**Comment by amro** [comment]
I'll release after I do some testing. So you know, the next few days are hectic for me. My guess is I'll have time Thursday.

**Comment by amro** [comment]
I just tried this and released version 1.1.6. Thanks @fsluis!
